cmd/orb_import.go

```diff
@@ -0,0 +1,204 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/CircleCI-Public/circleci-cli/api"
+	"github.com/CircleCI-Public/circleci-cli/api/graphql"
+)
+
+type orbImportPlan struct {
+	NewNamespaces           []string
+	NewOrbs                 []api.Orb
+	NewVersions             []api.OrbVersion
+	AlreadyExistingVersions []api.OrbVersion
+}
+
+func (o orbImportPlan) isEmpty() bool {
+	n := len(o.NewNamespaces)
+	n += len(o.NewOrbs)
+	n += len(o.NewVersions)
+	return n == 0
+}
+
+func importOrb(opts orbOptions) error {
+	vs, err := versionsToImport(opts)
+	if err != nil {
+		return err
+	}
+
+	plan, err := generateImportPlan(opts, vs)
+	if err != nil {
+		return err
+	}
+
+	displayPlan(os.Stdout, plan)
+	if !opts.noPrompt && !plan.isEmpty() && !opts.tty.askUserToConfirm("Are you sure you would like to proceed?") {
+		return nil
+	}
+
+	return applyPlan(opts, plan)
+}
+
+func versionsToImport(opts orbOptions) ([]api.OrbVersion, error) {
+	cloudClient := graphql.NewClient("https://circleci.com", "graphql-unstable", "", opts.cfg.Debug)
+
+	if opts.integrationTesting {
+		cloudClient = opts.cl
+	}
+
+	var orbVersions []api.OrbVersion
+	for _, ref := range opts.args {
+		if !isNamespace(ref) {
+			version, err := api.OrbInfo(cloudClient, ref)
+			if err != nil {
+				return nil, fmt.Errorf("orb info: %s", err.Error())
+			}
+
+			orbVersions = append(orbVersions, *version)
+			continue
+		}
+
+		// TODO: support an `--all-versions` flag that gets all versions instead of latest version per orb?
+		// Note: fetching all orb versions may not be possible. The best we could do is fetch an arbitrarily large number.
+		// Otherwise, do some other operation that grabs orb source data from a single namespace.
+		obv, err := api.ListNamespaceOrbVersions(cloudClient, ref)
+		if err != nil {
+			return nil, fmt.Errorf("list namespace orb versions: %s", err.Error())
+		}
+
+		orbVersions = append(orbVersions, obv...)
+	}
+
+	// Add a "short name" (the name of the orb without the namespace name,
```

Comment:
We actually went through a couple iterations of addressing the "shortname" issue, and this was the state that we disliked the least 🙃

I think `Orb.OrbName` would be confusing alongside `Orb.Name`, but I really like your idea of making it a method that just returns the string:
```go
o := Orb{
    Name: "circleci/node"
}
n := o.Shortname()
fmt.Println(n) // Prints 'circleci'.
```

It is technically less performant, since we will be parsing it out multiple times throughout the import flow but this seems negligible relative to all the network calls we'll be making!

Reply:
I think the annoying thing about `o.ShortName()` is that we weren't sure what to do about errors (e.g. we are passed `""` or `"foo"` instead of `"foo/bar"`)
- panic? but that seems not-idiomatic
- just return an empty string? but again, not idiomatic and could create a lot of confusion
- return an error? but that is frustrating number of additional lines of code to deal with a situation that should never happen
The more I think about it, the more I think we should just choose 1. or 2., since it's a situation that should literally never happen. But :question:

Reply:
@johnswanson After some thought, I actually feel that panic-ing as a result of a malformed orb name is an appropriate reaction.